nvflare/fuel/f3/streaming/byte_receiver.py

```diff
@@ -41,115 +41,244 @@
             if buf is None:
                 buf = bytes(0)
-
-            if 0 < chunk_size < len(buf):
-                result = buf[0:chunk_size]
-                # Put leftover to the head of the queue
-                self.task.buffers.appendleft((last_chunk, buf[chunk_size:]))
+            end_offset = self.chunk_offset + size
+            if 0 < end_offset < len(buf):
+                # Partial read
+                result = buf[self.chunk_offset : end_offset]
+                self.chunk_offset = end_offset
             else:
-                result = buf
+                # Whole chunk is consumed
+                if self.chunk_offset:
+                    result = buf[self.chunk_offset :]
+                else:
+                    result = buf
+
+                self.chunk_offset = 0
+                self.chunks.popleft()
+
                 if last_chunk:
-                    self.task.eos = True
+                    self.eos = True
 
-            self.task.offset += len(result)
+            self.offset += len(result)
 
-            if not self.task.last_chunk_received and (self.task.offset - self.task.offset_ack > self.ack_interval):
+            if not self.last_chunk_received and (self.offset - self.offset_ack > self.ack_interval):
                 # Send ACK
                 message = Message()
                 message.add_headers(
                     {
-                        StreamHeaderKey.STREAM_ID: self.task.sid,
+                        StreamHeaderKey.STREAM_ID: self.sid,
                         StreamHeaderKey.DATA_TYPE: StreamDataType.ACK,
-                        StreamHeaderKey.OFFSET: self.task.offset,
+                        StreamHeaderKey.OFFSET: self.offset,
                     }
                 )
-                self.byte_receiver.cell.fire_and_forget(STREAM_CHANNEL, STREAM_ACK_TOPIC, self.task.origin, message)
-                self.task.offset_ack = self.task.offset
+                self.cell.fire_and_forget(STREAM_CHANNEL, STREAM_ACK_TOPIC, self.origin, message)
+                self.offset_ack = self.offset
 
-            self.task.stream_future.set_progress(self.task.offset)
+            self.stream_future.set_progress(self.offset)
 
             return RESULT_DATA, result
 
+    def _append(self, buf: Tuple[bool, BytesAlike]):
+        if self.eos:
+            log.error(f"{self} Data after EOS is ignored")
+            return
+
+        self.chunks.append(buf)
+        self.next_seq += 1
+
+        # Wake up blocking read()
+        if not self.waiter.is_set():
+            self.waiter.set()
+
+
+class RxStream(Stream):
+    """A stream that's used to read streams from the streaming task"""
+
+    def __init__(self, task: RxTask):
+        super().__init__(task.size, task.headers)
+        self.task = task
+
+    def read(self, size: int) -> bytes:
+        if self.closed:
+            raise StreamError("Read from closed stream")
+
+        return self.task.read(size)
+
```

Comment:
what's the diff. RxTask VS RxStream? should we just merge RxTask into RxStream?

Reply:
RxStream is an implementation of Stream interface. RxTask is used to hike the details of a streaming task. Better to keep them separate.